runtime/src/transaction_batch.rs

```diff
@@ -46,12 +46,47 @@ impl<'a, 'b> TransactionBatch<'a, 'b> {
     pub fn needs_unlock(&self) -> bool {
         self.needs_unlock
     }
+
+    /// Release unnecessary locks before execution.
+    pub fn release_locks_early(&mut self, results: impl Iterator<Item = Result<()>>) {
+        // If we don't need to unlock, then we don't need to release locks early
+        if !self.needs_unlock() {
+            return;
+        }
+
+        let Self {
+            ref mut lock_results,
+            bank,
+            ref sanitized_txs,
+            ..
+        } = self;
+
+        let tx_and_is_locked = results
```

Comment:
anything we can do here to prevent user error as far as what's passed in to ensure results.len() == sanitized_txs.len() == lock_results.len()?

Reply:
we could pass results as a slice and do some length checks. I was trying to avoid an unneccessary allocation here: https://github.com/solana-labs/solana/pull/31823/files#r1213441822

Thinking now, I might be able to just make the result generic on Ok(T) so we don't need to add dep on `LoadedTransaction` in here.